mlir/lib/Catalyst/Driver/CompilerDriver.cpp

```diff
@@ -0,0 +1,443 @@
+    llvmModule->print(rawStringOstream, nullptr);
+    std::string outFile = fs::path("2_Enzyme.ll");
+    if (failed(catalyst::dumpToFile(options, outFile, outputs["Enzyme"]))) {
+        return failure();
+    }
+
+    return success();
+}
+
+LogicalResult runLowering(const CompilerOptions &options, MLIRContext *ctx, ModuleOp moduleOp,
+                          CompilerOutput::PipelineOutputs &outputs)
+{
+    auto pm = PassManager::on<ModuleOp>(ctx, PassManager::Nesting::Implicit);
+
+    std::unordered_map<const Pass *, std::list<Pipeline::Name>> pipelineTailMarkers;
+    for (const auto &pipeline : options.pipelinesCfg) {
+        if (failed(parsePassPipeline(joinPasses(pipeline.passes), pm, options.diagnosticStream))) {
+            return failure();
+        }
+        PassManager::pass_iterator p = pm.end();
+        const Pass *lastPass = &(*(p - 1));
+        pipelineTailMarkers[lastPass].push_back(pipeline.name);
+    }
+
+    if (options.keepIntermediate) {
+
+        {
+            std::string tmp;
+            {
+                llvm::raw_string_ostream s{tmp};
+                s << moduleOp;
+            }
+            std::string outFile = fs::path(options.moduleName.str()).replace_extension(".mlir");
+            if (failed(catalyst::dumpToFile(options, outFile, tmp))) {
+                return failure();
+            }
+        }
+
+        {
+            size_t pipelineIdx = 0;
+            auto printHandler =
+                [&](Pass *pass, CatalystIRPrinterConfig::PrintCallbackFn print) -> LogicalResult {
+                auto res = pipelineTailMarkers.find(pass);
+                if (res != pipelineTailMarkers.end()) {
+                    for (const auto &pn : res->second) {
+                        std::string outFile = fs::path(std::to_string(pipelineIdx++) + "_" + pn)
+                                                  .replace_extension(".mlir");
+                        {
+                            llvm::raw_string_ostream s{outputs[pn]};
+                            print(s);
+                        }
+                        if (failed(catalyst::dumpToFile(options, outFile, outputs[pn]))) {
+                            return failure();
+                        }
+                    }
+                }
+                return success();
+            };
+
+            pm.enableIRPrinting(std::unique_ptr<PassManager::IRPrinterConfig>(
+                new CatalystIRPrinterConfig(printHandler)));
+        }
+    }
+
+    if (failed(pm.run(moduleOp))) {
+        return failure();
+    }
+
+    return success();
+}
+
+LogicalResult QuantumDriverMain(const CompilerOptions &options, CompilerOutput &output)
+{
+    DialectRegistry registry;
+    static bool initialized = false;
+    if (!initialized) {
+        registerAllPasses();
+    }
+    initialized |= true;
+    registerAllCatalystPasses();
+    mhlo::registerAllMhloPasses();
+    gml_st::registerGmlStPasses();
```

Comment:
Is this associated with the scatter ticket? I cannot remember when/why we started registering `gml_st` in the past :thinking:

Reply:
In the past we also didn't need `mhlo::registerAllMhloPasses` because `quantum-opt` didn't handle any MHLO passes. Here we add `gml_st` passes because `scalarize` is a pass that we use. Otherwise, that pass is not available to us.